common/src/main/java/com/microsoft/identity/common/internal/ui/embeddedwebview/challengehandlers/PKeyAuthChallengeHandler.java

```diff
@@ -22,21 +22,259 @@
 // THE SOFTWARE.
 package com.microsoft.identity.common.internal.ui.embeddedwebview.challengehandlers;
 
-import android.os.Handler;
+import android.content.Intent;
+import android.support.annotation.NonNull;
+import android.webkit.WebView;
 
-import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationResponse;
+import com.microsoft.identity.common.adal.internal.AuthenticationConstants;
+import com.microsoft.identity.common.adal.internal.AuthenticationSettings;
+import com.microsoft.identity.common.adal.internal.IDeviceCertificate;
+import com.microsoft.identity.common.adal.internal.JWSBuilder;
+import com.microsoft.identity.common.adal.internal.util.StringExtensions;
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.exception.ErrorStrings;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.providers.oauth2.AuthorizationRequest;
 
-final class PKeyAuthChallengeHandler {
-    private Handler mRefHandler; //NOPMD
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.security.interfaces.RSAPrivateKey;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
 
-    private AuthorizationResponse mResponse; //NOPMD
+import static com.microsoft.identity.common.exception.ErrorStrings.DEVICE_CERTIFICATE_REQUEST_INVALID;
 
-    PKeyAuthChallengeHandler(Handler ref, AuthorizationResponse response) {
-        mRefHandler = ref;
-        mResponse = response;
+public final class PKeyAuthChallengeHandler implements IChallengeHandler {
+    private static final String TAG = PKeyAuthChallengeHandler.class.getSimpleName();
+    /**
+     * Certificate authorities are passed with delimiter.
+     */
+    private static final String CHALLENGE_REQUEST_CERT_AUTH_DELIMITER = ";";
+    private WebView mWebView;
+    private String mRedirectUri;
+    private AuthorizationRequest mRequest;
+    private IChallengeCompletionCallback mChallengeCallback;
+
+    /**
+     * @param redirectUri        Location: urn:http-auth:CertAuth?Nonce=<noncevalue>
+     *                           &CertAuthorities=<distinguished names of CAs>&Version=1.0
+     *                           &SubmitUrl=<URL to submit response>&Context=<server state that
+     *                           client must convey back>
+     * @param view
+     * @param completionCallback
+     * @throws ClientException
+     */
+    public PKeyAuthChallengeHandler(@NonNull final String redirectUri,

```

Comment:
This class really doesn't expose itself to be tested -- I think the `IChallengeHandler` interface needs some rethinking such that it provides some swappability/testability.

The `private` and `static` methods on this class do a TON of complex logic that we have no way to test.